core-bundle/src/Migration/AbstractRecordedMigration.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration;
+
+use Contao\CoreBundle\Entity\Migration as MigrationEntity;
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+use Symfony\Contracts\Service\ServiceSubscriberTrait;
+
+abstract class AbstractRecordedMigration extends AbstractMigration implements ServiceSubscriberInterface
+{
+    use ServiceSubscriberTrait;
+
+    public function shouldRun(): bool
+    {
+        return false === $this->hasRun();
+    }
+
+    protected function hasRun(): ?bool
+    {
+        if (!$this->connection()->getSchemaManager()->tablesExist(['tl_migration'])) {
+            return null;
+        }
+
+        return null !== $this->entityManager()->getRepository(MigrationEntity::class)->findOneBy(['name' => $this->getName()]);
+    }
+
+    protected function createResult(bool $successful, string $message = null): MigrationResult
+    {
+        if (false === $this->hasRun()) {
+            $migrationEntity = new MigrationEntity($this->getName());
+            $this->entityManager()->persist($migrationEntity);
+            $this->entityManager()->flush();
+        }
```

Comment:
```suggestion
        if (false === $this->hasRun()) {
            if (!$this->connection()->getSchemaManager()->tablesExist(['tl_migration'])) {
                $this->createMigrationTable();
            }
            $migrationEntity = new MigrationEntity($this->getName());
            $this->entityManager()->persist($migrationEntity);
            $this->entityManager()->flush();
        }
```

Reply:
I see, thank you, had no idea it was that simple 👍

Reply:
Or maybe something like this? 

```php
use Doctrine\ORM\Tools\SchemaTool;
…
private function createMigrationTable(): void
{
	$schemaTool = new SchemaTool($this->entityManager());
	$sql = $schemaTool->getCreateSchemaSql([
		$this->entityManager()->getClassMetadata(MigrationEntity::class),
	]);
	array_filter($sql, /* check for tl_migration */ );
	// Execute SQL...
}
```

Reply:
You probably also to create SQL from the schema and execute it.

Like  `$connection->exec($schemaTool->createSchema(…)->toSql($platform))`.

Reply:
@m-vo `createSchema` already executes it. The problem is that Contao's DoctrineSchemaListener adds all its tables via event and thus it fails.

Reply:
String operations on the SQL are a nasty hack, though. :smile: We could also clone the em and remove all event listeners (because you would not want any probably). But I feel there's a reason why doing this is so clunky. Maybe we should either just *not* do it or not use an entity to begin with and create the schema on the fly.

Reply:
Well, let's leave the nasty hack for now ;).